pkg/poll/poll.go

```diff
@@ -91,7 +91,11 @@ func WaitWithBackoffWithRetries(ctx context.Context, b backoff.Backoff, numRetri
 		sleep := b.Duration()
 		if deadline, ok := ctx.Deadline(); ok {
 			ctxSleep := time.Until(deadline)
-			sleep = minDuration(sleep, ctxSleep)
+			// We want to wait for smaller of backoff sleep and context sleep
```

Comment:
Question: IIUC from the PR description the `<-t.C` option is (often? always?) selected in the `select` block below when `sleep` is negative, is that the case?

If the deadline has passed, it seems that a sensible behavior would be to respect the context deadline by returning an error right away and thus not executing the operation.

It seems that with this change, `ctx.Done():` is going to be selected below when the context has expired, independent of how large `sleep` is, so in that case, why not simply make `sleep` positive (strictly greater than 0)? Would that be sufficient?

Reply:
Essentially this is a race condition. If `ctx` already expired but did not perfom the `cancel` yet, so there's nothing in the `ctx.Done()` channel. At the same time if we have negative timer, `t.C` will always have a value in the channel.
Just making sleep positive (in nanoseconds) doesn't really lets the context to finish in my tests.
We could just return a `context.DeadlineExceeded` error when `ctxSleep <= 0`, but then execution will continue in race with termination of the context. While technically correct (deadline did exceed), I don't know which consequences this might have.

Reply:
In the other hand additional 10ms of wait doesn't really add that much considering default backoff starts with 100ms and goes up from there (so context deadline in 10s of ms doesn't make much sense)